src/game.cpp

```diff
@@ -12316,15 +12316,23 @@ void game::fling_creature(Creature *c, const int &dir, float flvel, bool control
 
     // Fall down to the ground - always on the last reached tile
     if( !m.has_flag( "SWIMMABLE", c->pos() ) ) {
-        // Fall on ground
-        int force = rng( flvel, flvel * 2 ) / 9;
-        if( controlled ) {
-            force = std::max( force / 2 - 5, 0 );
-        }
-        if( force > 0 ) {
-            int dmg = c->impact( force, c->pos() );
-            // TODO: Make landing damage the floor
-            m.bash( c->pos(), dmg / 4, false, false, false );
+        const trap_id trap_under_creature = m.tr_at( c->pos() ).loadid;
+        // Didn't smash into a wall or a floor so only take the fall damage
+        if( thru && trap_under_creature == tr_ledge ) {
+            m.creature_on_trap( *c, false );
+        } else {
+            // Fall on ground
+            int force = rng( flvel, flvel * 2 ) / 9;
+            if( controlled ) {
+                force = std::max( force / 2 - 5, 0 );
+            }
+            if( force > 0 ) {
+                int dmg = c->impact( force, c->pos() );
+                // TODO: Make landing damage the floor
+                m.bash( c->pos(), dmg / 4, false, false, false );
+            }
+            // Always apply traps to creature i.e. bear traps, tele traps etc.
+            m.creature_on_trap( *c, false );
```

Comment:
This is going to cause creature_on_trap to fire twice if `thru == true && trap_under_creature == tr_ledge` no?

Reply:
Oh nevermind, I see now.  If it's a tr_ledge, ledge() will move them, then when the second creature_on_trap fires, it's checking if there's a trap on their new position.

But this begs another question, what happens if they are falling multiple z-levels?  I.e. what happens if the creature is falling 2 stories?  Obviously creature_on_trap with trigger twice as it should, but does the rest of it get handled gracefully as well?

Reply:
Hmm from testing in VS, the `m.creature_on_trap( *c, false );` should only fire once - either `thru && trap_under_creature == tr_ledge`, meaning you didn't hit a wall or floor and you're standing on a ledge trap (open space), or you hit something/are standing on floor/a trap and impact the ground.

I think that once you activate a ledge trap it then checks to make sure you're not standing on another ledge, but I'll have to test it.

Reply:
Right it should check if standing on another ledge in the second `m.creature_on_trap`, but then what happens if there's a third?

Reply:
It should only activate that once though, right? Since I have `if( thru && trap_under_creature == tr_ledge ) {`, and then `} else {`, either I activate `m.creature_on_trap( *c, false );` or I activate `// Fall on ground`...`m.creature_on_trap( *c, false );`. I don't think it activates both if and else in an if statement, does it?

Also just tested multiple z-levels (good idea, I hadn't thought of that!) and it worked as intended.

Reply:
I was mostly speculating through code inspection, if you tested it and everything seems to work well then we're good :+1: thanks!

Reply:
No problem, and thank you too! I would not have remembered to test multiple z-levels if you hadn't brought it up.

Reply:
> Since I have i if( thru && trap_under_creature == tr_ledge ) {, and then } else {, either I activate m.creature_on_trap( *c, false );

Wait, no, the second_ creature_on_trap is outside the if/else statement.  Is this what you intended?

Reply:
Wait I think you're being confused by the GitHub page. On line 12328, which is just below our comments, is where the if else ends. It probably looks like it's ending on line 12333 though since there's a nested if in there.

Reply:
That's exactly what happened, sorry for the thrash, I deleted my comment once I realized I had misread it.  Mea culpa.

Reply:
Oh no problem at all! If anything, I just thought I was misunderstanding there, haha!